### ISSUE TITLE
Peer dependency setup with working tests

### ISSUE THREAD

**Issue by crshmk** [ReactTraining/react-router#7521]
We're building a React component library. Some of our components use `Link`, with `BrowserRouter` installed in the parent app and listed as a dev dependency (which we connect locally using `npm-link`). 

parent setup looks like this:
```jsx
import { BrowserRouter as Router } from 'react-router-dom';
...
ReactDOM.render(
    <Router>
      <App />
    </Router>,
  app
);
```

imported component from the component library looks like this:
```jsx
 return (
    <Link
```

This works fine until we run tests in the component library, which needs access to `MemoryRouter`. So in the past we've just installed `react-router-dom` as a dev dependency in the component library, and that worked fine. 

After upgrading from version 4 to version 5, we get this runtime error:
```
Invariant failed: You should not use <Link> outside a <Router>
```

If I uninstall `react-router-dom` in our component library, there's no runtime error, but our component library tests cannot run as there's no access to `MemoryRouter`. 

I wonder if this error was added in v5, and if we might reconsider it. Alternatively, I wonder if we might export `MemoryRouter` as a standalone module. 

respective package.json files look like:

- parent app (where the BrowserRouter lives):
```json
 "dependencies": {
    "react-router-dom": "^5.2.0"
```

- component library (where we use a Link):

tests work; runtime error
```json
  "devDependencies": {
    "react-router-dom": "^5.2.0"
 ```

cannot run tests; no runtime error
```json
  "devDependencies": {}
```

  Note that in the past we've solved some npm-link and peer dependency issues by using explicit paths in the parent app's package.json moduleNameMapper, but this does not solve the issue:
```json
    "moduleNameMapper": {
      "^react-router-dom": "<rootDir>/node_modules/react-router-dom"
```

Thank you for any attention. 

react-router-dom version 5.2.0
node version 10.13.0
npm version 6.4.1

**Comment by timdorr** [comment]
This is because we use the Context API of React, which [creates a context instance](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/RouterContext.js#L11) that we pass around the library. It has to be the _exact same instance_, otherwise React can't match the Provider and Consumer pair.

This is a bit out of scope for the library, since it's not anything we have to solve on our end. But the basic idea is that you need the same instance of the `react-router-dom` module in both apps. Not sure what the best way is to do that, but you might try importing and re-exporting `react-router-dom` from your component library to ensure everyone is using the same copy of the module.

**Comment by crshmk** [comment]
Thanks much for the reply. We set about implementing your suggestion. The app runs and tests work when we 
```javascript
export {
  BrowserRouter,
  Link,
  Route,
  Router,
  Switch,
  useHistory,
  useLocation,
  withRouter
} from 'react-router-dom';
```
from the component library, but we'd have to ask all users of the component library to import all of your modules from the component library, which we felt was not ideal. Also, we found that tests in the parent app broke, and we thus had to also export a react-router dependency
```javascript
export { createMemoryHistory } from 'history';
```
from the component library, which seems like just the first example of why our general intuition felt this was not a good pattern. For now, we're going to revert to version 4 so the app will run again. We'll have to negotiate a better solution for the breaking change in the future. 

Thanks again for the help.
